### DISCUSSION TITLE
jhbuild libraries, Python pip module, etc

### DISCUSSION THREAD

**Discussion by totaam** [jralls/gtk-osx-build#83] [General]
First, you may want to consider switching to [libjpeg-turbo](https://libjpeg-turbo.org/) as **the** `libjpeg`, this commit does that: https://github.com/Xpra-org/gtk-osx-build/commit/b57ba2e9c7abc9a1ffcebeb4b5e4d34c8f9264b3
Every distribution that I know of, including [MSYS2](https://www.msys2.org/), Android, etc.. have done so. So there are no compatibility problems with any of the other libraries.
It's not just faster, it is also properly maintained - security updates, etc.
The only problem with that is `cmake` seems to use `rpath` in the library's self reference: https://github.com/Xpra-org/gtk-osx-build/issues/19
Is this a MacOS or jhbuild issue? Or should I report it to `libjpeg-turbo` instead?

---

I have a long list of Python modules that are installed using a jhbuild moduleset: https://github.com/Xpra-org/gtk-osx-build/blob/master/xpra-python3.modules
Some of the Python libraries no longer support `distutils`, and have to be installed using `pip`.
What would be the best way of doing this using `jhbuild`? Abusing `autogen`?

---

Doing full rebuilds, I hit some issues with `gtk-doc` requiring `pygments`, strangely enough this did not resolve the problem: https://github.com/Xpra-org/gtk-osx-build/commit/8ca590092ee06868eaa6b7b43a779b347af61e1a
(pygments would also need this: https://github.com/Xpra-org/gtk-osx-build/commit/96f36b302a7edb6106ffb3604c1821b5e54fa869)
Has anyone else seen that?

---

[I got bitten](https://github.com/Xpra-org/gtk-osx-build/commit/2cc47e14a6f5f5f84d07694c9a08c5f565247158) by the magic value `python3` which was missing from my list of `modules`... and then `jhbuild` ended up building things against the `pipenv` Python instead of the one I wanted.
Would you take a PR to add a note to the documentation at https://github.com/jralls/gtk-osx-build#python-configuration ?
(and also update the Python version on that page)

---

One more `jhbuild` question: how do I get `autotools` to run:
```shell
cd xar
./autogen.sh ....
```
[I tried](https://github.com/Xpra-org/gtk-osx-build/commit/824691bb71318c709b4d3c4e404a8bfa9220e82f): `<autotools id="xar" autogen-sh="cd xar;./autogen.sh" supports-non-srcdir-builds="no">` but that got mangled.
Surely there is an easier way of doing this?

**Comment by jralls** [comment]
A lot of questions!

Last one first, use `autogen-template` to customize the build commands for a module, see autotools in https://developer-old.gnome.org/jhbuild/stable/moduleset-syntax.html.en#moduleset-syntax-defs. But IIUC what's going on
```
    <autotools id="xar" supports-non-srcdir-builds="no" autogen-sh="autogen.sh">
```
should do that without needing an `autogen-template` attribute.

**Comment by jralls** [comment]
> Would you take a PR to add a note to the documentation at https://github.com/jralls/gtk-osx-build#python-configuration ?

It needs to be added to https://wiki.gnome.org/Projects/GTK/OSX/Building too. Getting edit on the Gnome wiki can be a pain so probably better if I do it.

**Comment by jralls** [comment]
> Some of the Python libraries no longer support distutils, and have to be installed using pip.
What would be the best way of doing this using jhbuild? Abusing autogen?

distutils is [deprecated in Python 3.10](https://peps.python.org/pep-0632/) and [removed in 3.12](https://github.com/python/cpython/issues/92584). The best way to fix it would be to create a setuptools module for jhbuild; although I'm a little leery of PyPI because of all of the [supply](https://www.hackread.com/pypi-packages-malware-supply-chain-attack/)[chain](https://www.itnews.com.au/news/pythons-pypi-registry-suffers-another-supply-chain-attack-589484)[attacks](https://arstechnica.com/information-technology/2022/09/actors-behind-pypi-supply-chain-attack-have-been-active-since-late-2021/) I guess a PIP module would also be a nice addition.

Failing (or pending) that I think cmake is easier to abuse than autotools. For example I've used it to [wrap boost-jam](https://github.com/Gnucash/gnucash-on-osx/blob/master/patches/boost-cmake.patch).

**Comment by jralls** [comment]
>First, you may want to consider switching to [libjpeg-turbo](https://libjpeg-turbo.org/) as the libjpeg, this commit does that: https://github.com/Xpra-org/gtk-osx-build/commit/b57ba2e9c7abc9a1ffcebeb4b5e4d34c8f9264b3
Every distribution that I know of, including [MSYS2](https://www.msys2.org/), Android, etc.. have done so. So there are no compatibility problems with any of the other libraries.
It's not just faster, it is also properly maintained - security updates, etc.
The only problem with that is cmake seems to use rpath in the library's self reference: https://github.com/Xpra-org/gtk-osx-build/issues/19
Is this a MacOS or jhbuild issue? Or should I report it to libjpeg-turbo instead?

PR welcome ;-)

As for rpath, I added [handling that](https://github.com/jralls/gtk-mac-bundler/commit/df5aa76f1c974d8a6508f52d8f22d051e1cb92cf) to bundler last spring. Did you have that in place when you experienced the problem?

To get cmake to set the install name you can try adding `-DCMAKE_MACOSX_RPATH=ON -DCMAKE_INSTALL_NAME_DIR=%{prefix}s` to `cmakeargs`; I'm not sure that `cmakeargs` understands the `%{prefix}s` substitution though so you may need to code it in jhbuildrc-custom using `cmakeargs-module`.
